lib/HLSL/DxilValidation.cpp

```diff
@@ -3988,6 +3988,114 @@ static void ValidateMetadata(ValidationContext &ValCtx) {
   }
 
   const hlsl::ShaderModel *SM = ValCtx.DxilMod.GetShaderModel();
+  // validate that any wavesize tags don't appear outside their expected shader
+  // models. Validate only 1 tag exists per entry point.
+  // Don't do this validation if the shader is non-compute
+  if (SM->IsCS() || SM->IsLib()) {
+
+    NamedMDNode *TA = pModule->getNamedMetadata("dx.entryPoints");
+    if (TA != nullptr) {
+      for (unsigned i = 0, end = TA->getNumOperands(); i < end; ++i) {
+        MDTuple *TANode = dyn_cast<MDTuple>(TA->getOperand(i));
+        if (TANode->getNumOperands() < 5) {
+          ValCtx.EmitMetaError(TANode, ValidationRule::MetaWellFormed);
+          return;
+        }
+        // get access to the digit that represents the metadata number that
+        // would store entry properties
+        const llvm::MDOperand &mOp =
+            TANode->getOperand(TANode->getNumOperands() - 1);
+        // the final operand to the entry points tuple should be a constant int.
```

Comment:
By the code, it looks like it should be a tuple not a constant int.

If the last operand isn't a tuple, should we just ignore it? Perhaps another well formed metadata error or at least an assert?

Reply:
Correct, I've adjusted the comment.
I believe the check is necessary on the final operand because there are some entry point functions that don't have entry point properties, and so the operand is null. In fact, that is how I decided to add that line of code, since I was working on an example that had an entry point without entry point properties. By adding an assert, 53 other tests will fail, so I will leave the continue; as is.